File: cpp/572. Subtree of Another Tree.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
// Time: O(mn)
public:
    bool isSameTree(TreeNode* s, TreeNode* t) {
        if(!s && !t) return true;
        if((s && !t) || (!s && t)) return false;
        return s->val==t->val && isSameTree(s->left, t->left) && isSameTree(s->right, t->right);
    }
    bool isSubtree(TreeNode* s, TreeNode* t) {
        if((s && !t) || (!s && t)) return isSameTree(s, t);
        return isSameTree(s, t) || isSubtree(s->left, t) || isSubtree(s->right, t);
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    
}

File: python/572-Subtree of Another Tree.py
from bintree_helper import *

class Solution:
    def isSubtree(self, s: TreeNode, t: TreeNode) -> bool:
        "Time: O(mn)"
        def isSameTree(s, t):
            if s is None and t is None: return True
            if s and t: return s.val==t.val and isSameTree(s.left, t.left) and isSameTree(s.right, t.right)
            return False
        if s is None or t is None: return isSameTree(s, t)
        return isSameTree(s, t) or self.isSubtree(s.left, t) or self.isSubtree(s.right, t)

s = deserialize("[3,4,5,1,2]")
t = deserialize("[4,1,2]")
ans = Solution().isSubtree(s, t)
assert ans==True

s = deserialize("[3,4,5,1,2,null,null,null,null,0]")
t = deserialize("[4,1,2]")
ans = Solution().isSubtree(s, t)
assert ans==False